code/controllers/subsystem/timer.dm

```diff
@@ -0,0 +1,518 @@
+			L.Insert(mid, src)
+
+		else
+			L += src
+		return
+
+	//get the list of buckets
+	var/list/bucket_list = SStimer.bucket_list
+
+	//calculate our place in the bucket list
+	var/bucket_pos = BUCKET_POS(src)
+
+	//get the bucket for our tick
+	var/datum/timedevent/bucket_head = bucket_list[bucket_pos]
+	SStimer.bucket_count++
+	//empty bucket, we will just add ourselves
+	if(!bucket_head)
+		bucket_list[bucket_pos] = src
+		return
+	//other wise, lets do a simplified linked list add.
+	if(!bucket_head.prev)
+		bucket_head.prev = bucket_head
+	next = bucket_head
+	prev = bucket_head.prev
+	next.prev = src
+	prev.next = src
+
+/datum/timedevent/Destroy()
+	..()
+	if(flags & TIMER_UNIQUE && hash)
+		SStimer.hashes -= hash
+
+	if(callBack && callBack.object && callBack.object != GLOBAL_PROC && callBack.object.active_timers)
+		callBack.object.active_timers -= src
+		UNSETEMPTY(callBack.object.active_timers)
+
+	callBack = null
+
+	if(flags & TIMER_STOPPABLE)
+		SStimer.timer_id_dict -= id
+
+	if(flags & TIMER_CLIENT_TIME)
+		if(!spent)
+			spent = world.time
+			SStimer.clienttime_timers -= src
+		return QDEL_HINT_IWILLGC
+
+	if(!spent)
+		spent = world.time
+		var/bucketpos = BUCKET_POS(src)
+		var/datum/timedevent/buckethead
+		var/list/bucket_list = SStimer.bucket_list
+		if(bucketpos > 0)
+			buckethead = bucket_list[bucketpos]
+
+		if(buckethead == src)
+			bucket_list[bucketpos] = next
+			SStimer.bucket_count--
+		else if(timeToRun < TIMER_MAX || next || prev)
+			SStimer.bucket_count--
+		else
+			var/l = length(SStimer.second_queue)
+			SStimer.second_queue -= src
+			if(l == length(SStimer.second_queue))
+				SStimer.bucket_count--
+
+		if(prev == next && next)
+			next.prev = null
+			prev.next = null
+		else
+			if(prev)
+				prev.next = next
+			if(next)
+				next.prev = prev
+	else
+		if(prev && prev.next == src)
+			prev.next = next
+		if(next && next.prev == src)
+			next.prev = prev
+	next = null
+	prev = null
+	return QDEL_HINT_IWILLGC
+
+/datum/timedevent/proc/getcallingtype()
+	. = "ERROR"
+	if(callBack.object == GLOBAL_PROC)
+		. = "GLOBAL_PROC"
+	else
+		. = "[callBack.object.type]"
+
+/proc/addtimer(datum/callback/callback, wait = 0, flags = 0)
+	if(!callback)
+		CRASH("addtimer called without a callback")
+
+	if(wait < 0)
+		stack_trace("addtimer called with a negative wait. Converting to 0")
+
+	//alot of things add short timers on themselves in their destroy, we ignore those cases
+	if(wait >= 1 && callback && callback.object && callback.object != GLOBAL_PROC && QDELETED(callback.object))
+		stack_trace("addtimer called with a callback assigned to a qdeleted object")
+
+	wait = max(wait, 0)
+
+	if(wait >= INFINITY)
+		CRASH("Attempted to create timer with INFINITY delay")
+
+	var/hash
+
+	if(flags & TIMER_UNIQUE)
+		var/list/hashlist
+		if(flags & TIMER_NO_HASH_WAIT)
+			hashlist = list(callback.object, "([callback.object.UID()])", callback.delegate, flags & TIMER_CLIENT_TIME)
```

Comment:
I do wonder if it's ok to use `UID()` over `\ref`

Reply:
What's the difference between the two?

Reply:
`UID` is a para unique thing to replace `\ref`; it can essentially be used the same, in most cases. It can be used for soft referencing (so you avoid GC issues). Right now, it's largely just UI's that enforce the useage of `UID`

Reply:
Thanks for clearing that up, much appreciated